File: frontend/src/lib/types/element.ts
/**
 * Core element type representing a visual object on the canvas.
 * Uses UUID primary keys to match Rails PostgreSQL schema.
 * `styles` is a flexible JSONB bag — extensible without migrations.
 */
export interface ElementStyles {
	backgroundColor?: string;
	borderRadius?: number;
	opacity?: number;
	borderWidth?: number;
	borderColor?: string;
}

export interface CanvasElement {
	id: string;
	canvas_id: string;
	name: string;
	element_type: string;
	x: number;
	y: number;
	width: number;
	height: number;
	rotation: number;
	z_index: number;
	styles: ElementStyles;
	locked: boolean;
	visible: boolean;
	version: number;
	created_at?: string;
	updated_at?: string;
}

/** Delta from the Go diff engine — field-level change tracking */
export interface ElementDelta {
	id: string;
	changes: Record<string, { from: unknown; to: unknown }>;
}

/** Broadcast event payload from Phoenix WebSocket */
export interface BroadcastEvent {
	event: 'element:updated' | 'element:created' | 'element:deleted' | 'element:batch_updated';
	payload: ElementDelta | CanvasElement | { ids: string[] } | { elements: CanvasElement[] };
}

File: api/app/models/element.rb
# Element model — a visual object on the canvas.
#
# Uses UUID primary keys for conflict-free distributed ID generation.
# `styles` is stored as JSONB — extensible without schema migrations.
# `version` enables optimistic concurrency control (Phase 6).
class Element < ApplicationRecord
  validates :canvas_id, presence: true
  validates :name, presence: true
  validates :element_type, presence: true
  validates :x, :y, :width, :height, :rotation, numericality: true
  validates :z_index, numericality: { only_integer: true }
  validates :version, numericality: { only_integer: true, greater_than: 0 }

  before_validation :set_defaults, on: :create

  scope :for_canvas, ->(canvas_id) { where(canvas_id: canvas_id).order(:z_index) }

  # Track previous state for diff engine notifications
  attr_accessor :previous_state

  after_update :capture_previous_state
  after_commit :notify_diff_engine, on: [:create, :update, :destroy]

  private

  def set_defaults
    self.canvas_id ||= "default"
    self.element_type ||= "rectangle"
    self.x ||= 0
    self.y ||= 0
    self.width ||= 100
    self.height ||= 100
    self.rotation ||= 0
    self.z_index ||= 0
    self.styles ||= {}
    self.locked ||= false
    self.visible = true if visible.nil?
    self.version ||= 1
    self.name ||= "Element #{z_index}"
  end

  def capture_previous_state
    # Stored by the controller before update
  end

  def notify_diff_engine
    DiffEngineNotifier.notify(self) if DiffEngineNotifier.available?
  rescue => e
    Rails.logger.warn "[DiffEngineNotifier] Failed: #{e.message}"
  end
end
